File: src/main/java/discretemath/structure/set/SimpleSet.java
package discretemath.structure.set;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Iterator;
import java.util.stream.Collectors;

import discretemath.structure.tuple.Pair;
import discretemath.structure.tuple.SimplePair;

public class SimpleSet<T> extends AbstractSet<T> {

	private final java.util.Set<T> set;

	private SimpleSet(java.util.Set<T> set) {
		this.set = set;
	}

	public static <T> Set<T> from(java.util.Set<T> set) {
		return new SimpleSet<>(java.util.Set.copyOf(set));
	}

	@SafeVarargs
	public static <T> Set<T> of(T... items) {
		return new SimpleSet<>(Arrays.stream(items).collect(Collectors.toUnmodifiableSet()));
	}

	@Override
	public int size() {
		return set.size();
	}

	@Override
	public boolean contains(T item) {
		return set.contains(item);
	}

	@Override
	public Set<T> unionWith(Set<T> set) {
		java.util.Set<T> newSet = new HashSet<>(this.set);

		if (set instanceof SimpleSet<T> simpleSet) {
			newSet.addAll(simpleSet.set);
		} else {
			set.forEach(newSet::add);
		}

		return new SimpleSet<>(newSet);
	}

	@Override
	public Set<T> intersectWith(Set<T> set) {
		java.util.Set<T> newSet = new HashSet<>(this.set);

		if (set instanceof SimpleSet<T> simpleSet) {
			newSet.retainAll(simpleSet.set);
		} else {
			newSet.removeIf(t -> !set.contains(t));
		}

		return new SimpleSet<>(newSet);
	}

	@Override
	public Set<T> differenceWith(Set<T> set) {
		java.util.Set<T> newSet = new HashSet<>(this.set);

		if (set instanceof SimpleSet<T> simpleSet) {
			newSet.removeAll(simpleSet.set);
		} else {
			newSet.removeIf(set::contains);
		}

		return new SimpleSet<>(newSet);
	}

	@Override
	public Set<T> symmetricDifferenceWith(Set<T> set) {
		return unionWith(set).differenceWith(intersectWith(set));
	}

	@Override
	public <P> Set<Pair<T, P>> cartesianProductWith(Set<P> set) {
		java.util.Set<Pair<T, P>> pairsSet = new HashSet<>(size() * set.size());

		for (T firstItem : this.set) {
			for (P secondItem : set) {
				Pair<T, P> pair = new SimplePair<>(firstItem, secondItem);
				pairsSet.add(pair);
			}
		}

		return new SimpleSet<>(pairsSet);
	}

	@Override
	public Iterator<T> iterator() {
		return set.iterator();
	}

	@Override
	public String toString() {
		return set.toString();
	}
}

File: src/main/scala/discretemath/structure/set/main/SimpleSet.scala
package discretemath.structure.set.main

import scala.collection.immutable


class SimpleSet[T] private(val set: immutable.Set[T]) extends AbstractSet[T] {

  override def cardinality: Int = set.size

  override def contains(item: T): Boolean = set.contains(item)

  override def unionWith[I <: Set[T], R >: Set[T]](set: I): R = SimpleSet(this.set.concat(set))

  override def intersectWith[I <: Set[T], R >: Set[T]](set: I): R = SimpleSet(this.set.view.filter(item => set.contains(item)).toSet)

  override def differenceWith[I <: Set[T], R >: Set[T]](set: I): R = SimpleSet(this.set.view.filter(item => !set.contains(item)).toSet)

  override def symmetricDifferenceWith[I <: Set[T], R >: Set[T]](set: I): R = unionWith(set).differenceWith(intersectWith(set))

  override def cartesianProductWith[P](set: Set[P]): Set[(T, P)] = SimpleSet(view.flatMap(x => set.map(y => (x, y))).toSet)

  override def powerSet: Set[Set[T]] = SimpleSet(set.subsets().map(subset => SimpleSet(subset)).toSet).asInstanceOf[Set[Set[T]]]

  override def iterator: Iterator[T] = set.iterator

  override def toString: String = set.toString
}

object SimpleSet {
  def apply[T](set: immutable.Set[T]): SimpleSet[T] = new SimpleSet[T](set)

  def apply[T](items: Iterable[T]): SimpleSet[T] = apply(items.toSet)

  def apply[T](items: T*): SimpleSet[T] = apply(immutable.Set(items: _*))
}
